### ISSUE TITLE
Class name must be a valid object or a string

### ISSUE THREAD

**Issue by evoalfio** [rezaamini-ir/migrator#41]
- PHP 8.2.4
- Laravel 10
- Release 1.5.7

```
[2023-04-20 13:58:11] local.ERROR: Class name must be a valid object or a string {"view":{"view":"/var/www/vs-workspace/pa/vendor/rezaamini-ir/migrator/resources/views/livewire/migration/read.blade.php","data":[]},"userId":119,"exception":"[object] (Spatie\\LaravelIgnition\\Exceptions\\ViewException(code: 0): Class name must be a valid object or a string at /var/www/vs-workspace/pa/vendor/rezaamini-ir/migrator/src/Service/MigratorParser.php:124)
[stacktrace]

#0 /var/www/vs-workspace/pa/vendor/rezaamini-ir/migrator/src/Http/Livewire/Migration/Single.php(75): Migrator\\Service\\MigratorParser->getPreview()

#1 /var/www/vs-workspace/pa/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Migrator\\Http\\Livewire\\Migration\\Single->mount()
```

I think it's due to the changes for the last release for the query preview functionality.

**Comment by Taoshan98** [comment]
Same error here when set 'local' option in config file to false

```log
Class name must be a valid object or a string {"view":{"view":"/var/www/html/mydomain.com/vendor/rezaamini-ir/migrator/resources/views/livewire/migration/read.blade.php","data":[]},"userId":112,"exception":"[object] (Spatie\\LaravelIgnition\\Exceptions\\ViewException(code: 0): Class name must be a valid object or a string at /var/www/html/mydomain.com/vendor/rezaamini-ir/migrator/src/Service/MigratorParser.php:124)
[stacktrace]
#0 /var/www/html/mydomain.com/vendor/rezaamini-ir/migrator/src/Http/Livewire/Migration/Single.php(75): Migrator\\Service\\MigratorParser->getPreview()
#1 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Migrator\\Http\\Livewire\\Migration\\Single->mount()
#2 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
#3 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\\Container\\Util::unwrapIfClosure()
#4 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(35): Illuminate\\Container\\BoundMethod::callBoundMethod()
#5 /var/www/html/mydomain.com/vendor/livewire/livewire/src/LifecycleManager.php(114): Illuminate\\Container\\BoundMethod::call()
#6 /var/www/html/mydomain.com/vendor/livewire/livewire/src/LivewireManager.php(130): Livewire\\LifecycleManager->mount()
#7 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php(338): Livewire\\LivewireManager->mount()
#8 /var/www/html/mydomain.com/vendor/rezaamini-ir/migrator/resources/views/livewire/migration/read.blade.php(34): Illuminate\\Support\\Facades\\Facade::__callStatic()
#9 /var/www/html/mydomain.com/vendor/livewire/livewire/src/ComponentConcerns/RendersLivewireComponents.php(83): include('...')
#10 /var/www/html/mydomain.com/vendor/livewire/livewire/src/ComponentConcerns/RendersLivewireComponents.php(84): Migrator\\Http\\Livewire\\Migration\\Read->Livewire\\ComponentConcerns\\{closure}()
#11 /var/www/html/mydomain.com/vendor/livewire/livewire/src/ComponentConcerns/RendersLivewireComponents.php(59): Livewire\\LivewireViewCompilerEngine->evaluatePath()
#12 /var/www/html/mydomain.com/vendor/sentry/sentry-laravel/src/Sentry/Laravel/Tracing/ViewEngineDecorator.php(45): Livewire\\LivewireViewCompilerEngine->get()
#13 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/View/View.php(195): Sentry\\Laravel\\Tracing\\ViewEngineDecorator->get()
#14 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/View/View.php(178): Illuminate\\View\\View->getContents()
#15 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/View/View.php(147): Illuminate\\View\\View->renderContents()
#16 /var/www/html/mydomain.com/vendor/livewire/livewire/src/Component.php(235): Illuminate\\View\\View->render()
#17 /var/www/html/mydomain.com/vendor/livewire/livewire/src/HydrationMiddleware/RenderView.php(14): Livewire\\Component->output()
#18 /var/www/html/mydomain.com/vendor/livewire/livewire/src/LifecycleManager.php(141): Livewire\\HydrationMiddleware\\RenderView::dehydrate()
#19 /var/www/html/mydomain.com/vendor/livewire/livewire/src/Macros/livewire-view-component.blade.php(3): Livewire\\LifecycleManager->initialDehydrate()
#20 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Filesystem/Filesystem.php(109): require('...')
#21 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Filesystem/Filesystem.php(110): Illuminate\\Filesystem\\Filesystem::Illuminate\\Filesystem\\{closure}()
#22 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/View/Engines/PhpEngine.php(58): Illuminate\\Filesystem\\Filesystem->getRequire()
#23 /var/www/html/mydomain.com/vendor/livewire/livewire/src/ComponentConcerns/RendersLivewireComponents.php(69): Illuminate\\View\\Engines\\PhpEngine->evaluatePath()
#24 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/View/Engines/CompilerEngine.php(70): Livewire\\LivewireViewCompilerEngine->evaluatePath()
#25 /var/www/html/mydomain.com/vendor/livewire/livewire/src/ComponentConcerns/RendersLivewireComponents.php(35): Illuminate\\View\\Engines\\CompilerEngine->get()
#26 /var/www/html/mydomain.com/vendor/sentry/sentry-laravel/src/Sentry/Laravel/Tracing/ViewEngineDecorator.php(45): Livewire\\LivewireViewCompilerEngine->get()
#27 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/View/View.php(195): Sentry\\Laravel\\Tracing\\ViewEngineDecorator->get()
#28 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/View/View.php(178): Illuminate\\View\\View->getContents()
#29 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/View/View.php(147): Illuminate\\View\\View->renderContents()
#30 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Http/Response.php(69): Illuminate\\View\\View->render()
#31 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Http/Response.php(35): Illuminate\\Http\\Response->setContent()
#32 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Routing/Router.php(906): Illuminate\\Http\\Response->__construct()
#33 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Routing/Router.php(875): Illuminate\\Routing\\Router::toResponse()
#34 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Routing/Router.php(797): Illuminate\\Routing\\Router->prepareResponse()
#35 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(141): Illuminate\\Routing\\Router->Illuminate\\Routing\\{closure}()
#36 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php(50): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#37 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Routing\\Middleware\\SubstituteBindings->handle()
#38 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Auth/Middleware/Authenticate.php(44): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#39 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Auth\\Middleware\\Authenticate->handle()
#40 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php(78): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#41 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\VerifyCsrfToken->handle()
#42 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php(49): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#43 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\View\\Middleware\\ShareErrorsFromSession->handle()
#44 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(121): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#45 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(64): Illuminate\\Session\\Middleware\\StartSession->handleStatefulRequest()
#46 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Session\\Middleware\\StartSession->handle()
#47 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php(37): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#48 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Cookie\\Middleware\\AddQueuedCookiesToResponse->handle()
#49 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php(67): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#50 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Cookie\\Middleware\\EncryptCookies->handle()
#51 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(116): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#52 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Routing/Router.php(797): Illuminate\\Pipeline\\Pipeline->then()
#53 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Routing/Router.php(776): Illuminate\\Routing\\Router->runRouteWithinStack()
#54 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Routing/Router.php(740): Illuminate\\Routing\\Router->runRoute()
#55 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Routing/Router.php(729): Illuminate\\Routing\\Router->dispatchToRoute()
#56 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(190): Illuminate\\Routing\\Router->dispatch()
#57 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(141): Illuminate\\Foundation\\Http\\Kernel->Illuminate\\Foundation\\Http\\{closure}()
#58 /var/www/html/mydomain.com/vendor/sentry/sentry-laravel/src/Sentry/Laravel/Http/SetRequestIpMiddleware.php(45): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#59 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Sentry\\Laravel\\Http\\SetRequestIpMiddleware->handle()
#60 /var/www/html/mydomain.com/vendor/sentry/sentry-laravel/src/Sentry/Laravel/Http/SetRequestMiddleware.php(30): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#61 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Sentry\\Laravel\\Http\\SetRequestMiddleware->handle()
#62 /var/www/html/mydomain.com/vendor/livewire/livewire/src/DisableBrowserCache.php(19): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#63 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Livewire\\DisableBrowserCache->handle()
#64 /var/www/html/mydomain.com/vendor/barryvdh/laravel-debugbar/src/Middleware/InjectDebugbar.php(66): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#65 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Barryvdh\\Debugbar\\Middleware\\InjectDebugbar->handle()
#66 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#67 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php(31): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle()
#68 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\ConvertEmptyStringsToNull->handle()
#69 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#70 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php(40): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle()
#71 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\TrimStrings->handle()
#72 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#73 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize->handle()
#74 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php(86): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#75 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\PreventRequestsDuringMaintenance->handle()
#76 /var/www/html/mydomain.com/vendor/fruitcake/laravel-cors/src/HandleCors.php(38): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#77 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Fruitcake\\Cors\\HandleCors->handle()
#78 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Http/Middleware/TrustProxies.php(39): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#79 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Http\\Middleware\\TrustProxies->handle()
#80 /var/www/html/mydomain.com/vendor/sentry/sentry-laravel/src/Sentry/Laravel/Tracing/Middleware.php(52): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#81 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Sentry\\Laravel\\Tracing\\Middleware->handle()
#82 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(116): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#83 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(165): Illuminate\\Pipeline\\Pipeline->then()
#84 /var/www/html/mydomain.com/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(134): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter()
#85 /var/www/html/mydomain.com/public/index.php(52): Illuminate\\Foundation\\Http\\Kernel->handle()
#86 {main}
```

**Comment by dokko-dari** [comment]
> * PHP 8.2.4
> * Laravel 10
> * Release 1.5.7
> 
> ```
> [2023-04-20 13:58:11] local.ERROR: Class name must be a valid object or a string {"view":{"view":"/var/www/vs-workspace/pa/vendor/rezaamini-ir/migrator/resources/views/livewire/migration/read.blade.php","data":[]},"userId":119,"exception":"[object] (Spatie\\LaravelIgnition\\Exceptions\\ViewException(code: 0): Class name must be a valid object or a string at /var/www/vs-workspace/pa/vendor/rezaamini-ir/migrator/src/Service/MigratorParser.php:124)
> [stacktrace]
> 
> #0 /var/www/vs-workspace/pa/vendor/rezaamini-ir/migrator/src/Http/Livewire/Migration/Single.php(75): Migrator\\Service\\MigratorParser->getPreview()
> 
> #1 /var/www/vs-workspace/pa/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Migrator\\Http\\Livewire\\Migration\\Single->mount()
> ```
> 
> I think it's due to the changes for the last release for the query preview functionality.


I had a same error. and I found problems in my migration files.

for Migrator, migration file have to start like below

```
return new class extends Migration
{
```

not like

```
class SomeTable extends Migration 
{
```

if install some packages, can have publishing migration file like this.
Hope this helps you :)

**Comment by reziamini** [comment]
This error comes when you are working with some old migrations that the migrator does not support right now (old-style migrations).
In the new version `1.5.8`, we are skipping these migrations from the preview.
Please check this version out and tell me the result.

**Comment by evoalfio** [comment]
@rezaamini-ir 

i have upgraded to `1.5.8` version but got same error: 

> Class name must be a valid object or a string.

actually i have three migrations, two of them start with 

`class SomeTable extends Migration 
{`

and one starts with:

`return new class extends Migration {`

it seems that the two migrations not are skipped during reading list 

Thanks to all for support

**Comment by evoalfio** [comment]
@rezaamini-ir 

regarding the first error i resolve it only if i rename my first two migrations to: 

> return new class extends Migration {

regarding the second error you should bypass all files that doesn't end with *.php, as for example the .gitkeep file

Hope i can help :-)

**Comment by Taoshan98** [comment]
PHP 8.2.6
Laravel 9
Release 1.5.8

I also updated but I encounter the same error, I tried first to edit all my migrations like this:
 ```php
return new class extends Migration {
```
but it doesn't work.

Then I tried to delete all migration files, keeping a backup clearly, 
but it doesn't work.

Then I tried to create an empty migration, without any operation inside,
but it doesn't work.

There are the logs.
```
2023-06-13 08:34:18] local.ERROR: Class name must be a valid object or a string {"view":{"view":"/var/www/html/mydomain.com/vendor/rezaamini-ir/migrator/resources/views/livewire/migration/read.blade.php","data":[]},"userId":112,"exception":"[object] (Spatie\\LaravelIgnition\\Exceptions\\ViewException(code: 0): Class name must be a valid object or a string at /var/www/html/mydomain.com/vendor/rezaamini-ir/migrator/src/Service/MigratorParser.php:125)
[stacktrace]
#0 /var/www/html/mydomain.com/vendor/rezaamini-ir/migrator/src/Http/Livewire/Migration/Single.php(75): Migrator\\Service\\MigratorParser->getPreview()
```

**Comment by evoalfio** [comment]
Have you tried to remove all other files that you have in the folder containing the migrations?
For example i had to remove an incorrect gitkeep file (without the '.' prefix) that cause me the error.

**Comment by Taoshan98** [comment]
I solved it this way: 
give permissions 777 to the /database folder (strange) however maybe it is because using also a sqlite db I had to give the ownership of the /database folder to $USER:www-data.

Then I modified the package code like this:
```php

//$migrationObject = new(resolve(Filesystem::class)->getRequire($migration));

$filesystem = resolve(Filesystem::class);
$migrationObject = $filesystem->getRequire($migration);

if (is_int($migrationObject)){
    return [];
}
```
So the package started working, I don't know if it is right as a solution.

I also noticed that if I set the 'per_page' property in the configuration file to a value greater than 30 the application dies,
also happens if I change page in migration table.
